### PR TITLE
Bitpackingv2

### PR DESCRIPTION
**Pull request by vayuda**
Improved code structure of the pack/unpack functions
Now supports trinary for bitnet applications
Now supports packing along any dimension for any size tensors
Now supports packing/unpacking elements such that lower indexed elements can be placed in the highest order bits or lowest order bits in the container

### CODE REVIEW

File: test/prototype/test_bitpacking.py [hunk 1]
```diff
@@ -1,70 +1,45 @@
...
-    packed = pack(test_tensor,16, 3)
-    unpacked = unpack(packed, 3)
-    unpadded = unpacked[:test_tensor.shape[0], ...]
-    assert(unpadded.allclose(test_tensor))
-
-@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
-@pytest.mark.skipif(not has_triton(), reason="unsupported without triton")
-def test_uint2_to_uint8_col_wise_compile():
-    torch._dynamo.config.specialize_int = True
-    pack_compiled = torch.compile(pack, fullgraph=True)
```
Comment (msaroufim) [inline test/prototype/test_bitpacking.py:56]: I noticed you're not using compile anymore? I also worry that functions that include `.shape` in them will cause a cuda sync and be very slow

File: torchao/prototype/common/bitpacking.py [hunk 2]
```diff
@@ -1,101 +1,125 @@
...
-from functools import reduce
+from typing import Optional, Union
 
+def mod_shape(shape, mod, dim):
+    """changes a select dimension of the input shape to mod"""
+    return (*shape[:dim], mod, *shape[dim+1:])
 
 
-def unpack(data, data_size, by_rows = True, device="cuda"):
+def dtype_to_bits(dtype):
```
Comment (msaroufim) [inline torchao/prototype/common/bitpacking.py:9]: nit: can you make this a dictionary instead?

File: test/prototype/test_bitpacking.py [hunk 3]
```diff
@@ -1,70 +1,45 @@
...
 
-def test_uint3_to_int16_col_wise_cpu():
-    test_tensor = torch.randint(0, 7, (8, 5), dtype=torch.int16)
-    packed = pack(test_tensor,16, 3, False, device='cpu')
-    unpacked = unpack(packed, 3, False, device='cpu')
-    unpadded = unpacked[:test_tensor.shape[0], ...]
-    assert(unpadded.allclose(test_tensor))
-    
+def test_to_uint8_CPU():
+    for dtype in {torch.uint2, torch.uint3, torch.uint4, torch.uint5, torch.uint6, torch.uint7}:
```
Comment (msaroufim) [inline test/prototype/test_bitpacking.py:18]: favor pytest parametrize that way it'll be easier to identify the failing test in CI

**Review by msaroufim — CHANGES_REQUESTED** [review]
Left some feedback here and in cuda mode, my worry is this code will be slow so will need some benchmarks and it's a big loss to lose torch.compile support

File: test/prototype/test_bitpacking.py [hunk 4]
```diff
@@ -1,70 +1,84 @@
...
-    test_tensor = torch.randint(0, 15, (3, 4), dtype=torch.uint8).cuda()
-    packed = pack_compiled(test_tensor, 8, 4)
-    unpacked = unpack_compiled(packed, 4)
-    unpadded = unpacked[:test_tensor.shape[0], ...]
-    assert(unpadded.allclose(test_tensor))
+@pytest.mark.parametrize("dtype", dtypes)
+@pytest.mark.parametrize("dim", dimensions)
+def test_compile(dtype, dim):
+    pack_compile = torch.compile(pack, fullgraph=True)
+    unpack_compile = torch.compile(unpack, fullgraph=True)
```
Comment (msaroufim) [inline test/prototype/test_bitpacking.py:50]: you need to run the function to test it since compile is a JIT compiler

File: torchao/prototype/common/bitpacking.py [hunk 5]
```diff
@@ -1,101 +1,139 @@
...
+    **Pads rows to be divisible by the scale**
+    TODO: support something like packing 8 uint 3s into 3 uint8s
     
     Inputs:
-    data: torch.Tensor - a tensor of unpacked elements of a small dtype.
-    container_size: int - the size of the large dtype in bits.
-    data_size: int - the size of the small dtype in bits.
+    data: a tensor of unpacked elements of a small dtype. The dtype used for the data will be used for the container.
+    dimension: the dimension to pack along
+    element_dtype: the dtype of the elements to pack
```
Comment (gau-nernst) [inline torchao/prototype/common/bitpacking.py:88]: I think keeping `element_size` (or maybe change to `element_bit_width`) is better. Then `dtype_to_bits()` doesn't need to support uint2-7 (so you can use `torch.finfo()` and `torch.iinfo()`)

File: torchao/prototype/common/bitpacking.py [hunk 6]
```diff
@@ -1,101 +1,133 @@
...
-    return torch.stack(unpacked_data,dim=-1).view(*shape[:-1],shape[-1]*scale) # stack the unpacked data and reshape to the original shape
+        slices[dim] = slice(i, None, scale)
+        if order:
+            packed |= data[slices] << container_size-element_bit_width*(i+1)
+        else:
+            packed |= data[slices] << element_bit_width*i
+    return packed
 
-def _pack_by_rows(data, container_size, data_size, device) -> torch.Tensor:
+if __name__ == '__main__':
```
Comment (msaroufim) [inline torchao/prototype/common/bitpacking.py:110]: Not a huge fan of having usage instructions in a python file. Ideally this should be a test instead or some usage example in a README in the `common/` folder

File: torchao/prototype/common/bitpacking.py [hunk 7]
```diff
@@ -1,101 +1,133 @@
...
-    return ret.view(shape[0] // scale, *shape[1:]).to(device)
-
-def _pack_by_cols(data, container_size, data_size, device) -> torch.Tensor:
-    scale = container_size // data_size
-    assert scale > 1, f"container_size ({container_size}) not double the capacity ofdata_size ({data_size})"
-    # pad the data to be divisible by scale
-    if data.shape[-1] % scale != 0:
-        padding = torch.zeros((*data.shape[:-1], scale - data.shape[-1] % scale), dtype=data.dtype).to(device)
-        data = torch.cat([data, padding], dim=-1).cuda()
+    if element_type == "trinary":
```
Comment (msaroufim) [inline torchao/prototype/common/bitpacking.py:120]: can you add a quick comment here?

File: torchao/prototype/common/bitpacking.py [hunk 8]
```diff
@@ -1,101 +1,133 @@
...
+        test_tensor = torch.randint(-1, 1, shape, dtype=torch.int8).cuda()
+    else:
+        test_tensor = torch.randint(0, 2**element_bit_width, shape, dtype=torch.uint8).cuda()
+        
+    packed = pack_compile(test_tensor, element_bit_width, element_type=element_type, dim = dim, container_dtype = torch.uint8, pad= True)
+    print(packed.shape)
```
Comment (msaroufim) [inline torchao/prototype/common/bitpacking.py:126]: turn these into logging statements

**Review by msaroufim — CHANGES_REQUESTED** [review]
few minor pieces of feedback: comments would be nice too

**Review by msaroufim — APPROVED** [review]
